### PR TITLE
Fix async error handler

### PR DESCRIPTION
**Pull request by Gcaufy**
## I'm submitting a...

```
[x] Bug Fix
[ ] Feature
[ ] Other (Refactoring, Added tests, Documentation, ...)
```

## Checklist

- [x] Commit Messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
- [x] Tests for the changes have been added

## Description

https://github.com/wechaty/wechaty/blob/master/src/wechaty.ts#L618

This code doesn't handle async errors, which means they user have to write code like this, other wise the error will goes to unhandled promises.

```
bot.on('message', (m) => {
  try {
    // do all my works
  } catch(e) {
   // handle all my errors
  }
})
```

## Does this PR introduce a breaking change?

```
[x] Yes
[ ] No
```

This is going t break thing. the user's `emit('error')` may get more errors than before.

### CODE REVIEW

File: src/wechaty.ts [hunk 1]
```diff
@@ -610,15 +610,22 @@ export class Wechaty extends EventEmitter implements Sayable {
   private addListenerFunction (event: WechatyEventName, listener: AnyFunction): void {
     log.verbose('Wechaty', 'addListenerFunction(%s)', event)
 
+    const handleError = (e: Error, type = '') => {
+      log.error('Wechaty', 'addListenerFunction(%s) listener %sexception: %s', event, type ? '' : (type + ''), e)
```
Comment (windmemory) [inline src/wechaty.ts:614]: Seems like the `type` is a string and will always be a string, so I don't think we need to do `type ? '' : (type + '')` here, which is redundant.

```suggestion
      log.error('Wechaty', 'addListenerFunction(%s) listener %s exception: %s', event, type, e)
```

File: src/wechaty.ts [hunk 2]
```diff
@@ -610,15 +610,22 @@ export class Wechaty extends EventEmitter implements Sayable {
...
+      this.emit('error', e)
+    }
+
     /**
      * We use `super.on()` at here to prevent loop
      */
     super.on(event, (...args: any[]) => {
       try {
-        listener.apply(this, args)
+        const rst = listener.apply(this, args)
```
Comment (windmemory) [inline src/wechaty.ts:623]: Is `rst` stands for `result`? If so, I would prefer to use `result` here.

**Comment by huan** [comment]
@Gcaufy  Thank you very much for adding the async error support for Wechaty, I believe it's necessary to catch the async error for our system and emit an `error` event to prevent uncaught promise rejection.

@windmemory Thanks for the code review, please help @Gcaufy to get this PR approved from you, then we will be able to merge it.

Wechaty benefited lots of the efforts from you for the good, I really appreciate it!

File: src/wechaty.ts (hunk 1 above)
Comment (Gcaufy) [reply src/wechaty.ts]: Have no idea why I did it like this. lol

**Comment by windmemory** [comment]
@huan This PR looks good to me, approved already!

**Comment by huan** [comment]
Thank you very much for enhancing the error handler for our listeners @Gcaufy @windmemory, appreciate it!

**Pull request merged**